### PR TITLE
Use coordinator + worker libs

### PR DESCRIPTION
**Pull request by michaeldmitry**
## Context 
Implement generic worker lib in `cos-lib`.

## Testing Consideration
`tox -e unit`
`tox -e scenario`

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -9,250 +9,61 @@
...
-                )
-            )
-        else:
+        self.worker._on_collect_status(e)
+        # add Tempo worker custom blocking conditions
+        if self.worker.roles and self.worker.roles[0] not in self._valid_roles:
             logger.error(
                 f"`role` config value {self.config.get('role')!r} invalid: should "
-                f"be one of {list(TempoRole)}."
+                f"be one of {(self._valid_roles)}."
```
Comment (sed-i) [inline src/charm.py:66]: This will be logged every hook. Is that desirable?

File: src/charm.py [hunk 2]
```diff
@@ -261,12 +72,37 @@ def _on_collect_status(self, e: CollectStatusEvent):
...
+
+    def pebble_layer(self, worker: Worker) -> Layer:
+        """Return a dictionary representing a Pebble layer.
+
+        Caller is responsible for checking whether the tempo role is valid before
+        calling this method.
+        """
+        role = "".join(worker.roles)
+        if role == "all":
+            role = "scalable-single-binary"
```
Comment (sed-i) [inline src/charm.py:91]: Not sure, but should this be:
```suggestion
        role = "scalable-single-binary" if "all" in worker.roles else "".join(worker.roles)
```

File: src/charm.py [hunk 3]
```diff
@@ -261,12 +72,37 @@ def _on_collect_status(self, e: CollectStatusEvent):
...
+        return Layer(
+            {
+                "summary": "tempo worker layer",
+                "description": "pebble config layer for tempo worker",
+                "services": {
+                    "tempo": {
+                        "override": "replace",
+                        "summary": "tempo worker process",
+                        "command": f"/bin/tempo -config.file={CONFIG_FILE} -target {role}",
+                        "startup": "enabled",
```
Comment (sed-i) [inline src/charm.py:101]: Iiuc, we need to have some config in place to start tempo correctly, so not sure if "startup enabled" is appropriate.

File: src/charm.py (hunk 1 above)
Comment (michaeldmitry) [reply src/charm.py:66]: following the pattern in shared worker's `collect_unit_status`, maybe we don't actually. Setting the status would be enough I guess.

File: src/charm.py (hunk 3 above)
Comment (michaeldmitry) [reply src/charm.py:94]: Yes. So, I've added a safeguard in the shared worker to not start/restart the pebble service if the config doesn't exist yet. Once the file gets created, I believe we have a guarantee that during the workload lifecycle, we'll always have that file (no other service deletes it anywhere).

**Review by PietroPasotti — APPROVED** [review]

**Pull request merged**